File: Day6/Day6.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <cstring>

long array_sum(long* array, int size) {
    long sum = 0;
    for(int i = 0; i <= size; i++) {
        sum += (long) array[i];
    }
    return sum;
}

long simulate(std::string line, int n_days) {
    long generation[9] =  {0};
    for(std::string::iterator it = line.begin(); it != line.end(); it++) {
        if(*it != ',') {
            generation[(long) *it - '0']++;
        } 
    }
    for(int i = 0; i < n_days; i++) {
        long tmp = generation[0];
        for(int j = 1; j <= 8; j++) {
            generation[j-1] += generation[j];
            generation[j] = 0;
        }
        generation[0] -= tmp;
        generation[6] += tmp;
        generation[8] += tmp;
    }

    return array_sum(generation, 8);
}

int main(int argc, char* argv[]) {
    std::ifstream file(argv[1]);
    std::string line("");
    std::getline(file, line);  
    std::cout << "Part1: " << simulate(line, 80) << std::endl;
    std::cout << "Part2: " << simulate(line, 256) << std::endl;
    file.close();
    return 0;
}

File: Day6/Day6.py
import sys

def simulate(lines, n_days):
    day = 0
    generation = [0, 0, 0, 0, 0, 0, 0, 0, 0]

    for value in lines.split(","):
        number = int(value)
        generation[number] += 1
    for i in range(1, n_days+1):
        temp = list(generation)
        for j in range(1, 9):
            generation[j-1] += generation[j]
            generation[j] = 0
        generation[0] -= temp[0]
        generation[6] += temp[0]
        generation[8] += temp[0]
    
    return sum(generation)




if __name__ == "__main__":
    file = open(sys.argv[1], "r")
    file_lines = file.readline()
    print("Part1: ", simulate(file_lines, 80))
    print("Part2: ", simulate(file_lines, 256))
    file.close()
